Python/pylifecycle.c

```diff
@@ -1646,6 +1649,42 @@ Py_FinalizeEx(void)
 
     /* Get current thread state and interpreter pointer */
     PyThreadState *tstate = _PyRuntimeState_GetThreadState(runtime);
+    PyInterpreterState *interp = tstate->interp;
+
+    /* Check we're running in the main interpreter (not yet supported to call
+     * from any interpreter).
+     */
+    if (interp != PyInterpreterState_Main()) {
+        Py_FatalError("must be called from the main interpreter\n");
+    }
+
+    // Finalize sub-interpreters.
+    PyThread_acquire_lock(runtime->interpreters.mutex, WAIT_LOCK);
+    runtime->interpreters.allow_new = 0;
+    PyThread_release_lock(runtime->interpreters.mutex);
+    PyInterpreterState *curr_interp = PyInterpreterState_Head();
+    PyInterpreterState *next_interp;
+    int64_t num_destroyed = 0;
+    while (curr_interp != NULL) {
+        next_interp = PyInterpreterState_Next(curr_interp);
+        if (curr_interp != interp) {
+            PyThreadState_Swap(curr_interp->tstate_head);
+            Py_EndInterpreter(curr_interp->tstate_head);
+            num_destroyed++;
+        }
+        curr_interp = next_interp;
+    }
+    PyThreadState_Swap(tstate);
+
+    if (num_destroyed > 0) {
+        /* Sub-interpreters were still running, but should have be finalized
+         * before finalizing the runtime.
+         */
+        if (PyErr_ResourceWarning(NULL, 1,
+                                  "extra %zd interpreters", num_destroyed)) {
```

Comment:
I'm not sure why, but for some reason this warning isn't being output in the tests on Windows only. Anyone have any ideas?